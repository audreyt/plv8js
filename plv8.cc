#include "plv8.h"

#include <v8.h>

#ifdef __cplusplus
extern "C"{
#endif

#include "funcapi.h"
#include "catalog/pg_proc.h"
#include "catalog/pg_type.h"
#include "mb/pg_wchar.h"
#include "utils/array.h"
#include "utils/memutils.h"
#include "utils/builtins.h"
#include "utils/lsyscache.h"
#include "utils/syscache.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(plv8_call_handler);
PG_FUNCTION_INFO_V1(plv8_call_validator);

void		_PG_init(void);

#ifdef __cplusplus
} // extern "C"
#endif

typedef struct plv8_type_info
{
	Oid			typid;
	Oid			ioparam;
	int32		typmod;
	int16		len;
	bool		byval;
	char		align;
	char		category;
	FmgrInfo	fn_input;
	FmgrInfo	fn_output;

	/* optional, only when category = TYPECATEGORY_ARRAY */
	struct plv8_type_info *elem;
} plv8_type_info;

typedef struct
{
	v8::Persistent<v8::Function>	function;
	char	   *proname;
	char	   *prosrc;

	TransactionId fn_xmin;
	ItemPointerData fn_tid;

	int			nargs;
	plv8_type_info		rettype;
	plv8_type_info		argtypes[FUNC_MAX_ARGS];
/*
	Oid			rettype;
	Oid			retioparam;
	int32		rettypmod;
	char		retcategory;
	FmgrInfo   *fn_retinput;
	int			nargs;
	Oid			argtypes[FUNC_MAX_ARGS];
	char		argcategories[FUNC_MAX_ARGS];
	/ * for array type arguments * /
	Oid			elemtypes[FUNC_MAX_ARGS];
	int16		elemlens[FUNC_MAX_ARGS];
	bool		elembyvals[FUNC_MAX_ARGS];
	char		elemaligns[FUNC_MAX_ARGS];
	
	FmgrInfo   *fn_argoutput[FUNC_MAX_ARGS];
*/
} plv8_info;

typedef struct
{
	char		proc_name[NAMEDATALEN];
	plv8_info  *info_data;
} plv8_info_entry;


v8::Persistent<v8::Context> global_context = v8::Context::New();
v8::TryCatch global_try_catch;
static HTAB *plv8_info_hash = NULL;


//static Datum plv8_func_handler(PG_FUNCTION_ARGS);
static plv8_info *compile_plv8_function(Oid fn_oid, bool is_trigger);

static v8::Persistent<v8::Function> plv8_create_function(const char *proname, int proarglen,
														 const char *proargs[], const char *prosrc);
static Datum plv8_call_jsfunction(FunctionCallInfo fcinfo);
static void plv8_fill_type(Oid typid, plv8_type_info *type, bool forinput, bool output);
static Datum plv8_jsarray_to_datum(v8::Handle<v8::Value> value,
								   bool *isnull, Oid elemtype, Oid elemioparam, int16 elemlen,
								   bool elembyval, char elemalign, FmgrInfo *fn_eleminput);
static Datum plv8_jsval_to_datum(v8::Handle<v8::Value> value,
								 bool *isnull, Oid typid, Oid ioparam, FmgrInfo *fn_input);
static v8::Handle<v8::Value> plv8_datum_to_jsarray(Datum datum,
						 bool isnull, Oid elemtype, int16 elemlen,
						 bool elembyval, char elemalign, FmgrInfo *fn_elemoutput);
static v8::Handle<v8::Value> plv8_datum_to_jsval(Datum datum, bool isnull, Oid type, FmgrInfo *fn_output);
static char *plv8_toutf(char *cstr);
static char *plv8_tolocal(char *cstr);
static void plv8_elog_exception(int errcode, v8::TryCatch *try_catch);
static const char *ToCString(v8::String::Utf8Value &value);


void
_PG_init(void)
{
	static bool inited = false;
	HASHCTL		hash_ctl;

	if (inited)
		return;

	hash_ctl.keysize = NAMEDATALEN;
	hash_ctl.entrysize = sizeof(plv8_info);

	plv8_info_hash = hash_create("PLv8 Procedures",
								 32,
								 &hash_ctl,
								 HASH_ELEM);

	inited = true;
}

Datum
plv8_call_handler(PG_FUNCTION_ARGS)
{
	Oid			funcOid = fcinfo->flinfo->fn_oid;

	if (!fcinfo->flinfo->fn_extra)
	{
		fcinfo->flinfo->fn_extra = (void *) compile_plv8_function(funcOid, false);
	}

	PG_RETURN_DATUM(plv8_call_jsfunction(fcinfo));
}

Datum
plv8_call_validator(PG_FUNCTION_ARGS)
{
	Oid			funcOid = PG_GETARG_OID(0);
	HeapTuple	tuple;
	Form_pg_proc proc;
	char		functyptype;
	int			numargs;
	Oid		   *argtypes;
	char	  **argnames;
	char	   *argmodes;
	bool		istrigger = false;
	int			i;

	/* Get the new function's pg_proc entry */
	tuple = SearchSysCache(PROCOID,
						   ObjectIdGetDatum(funcOid),
						   0, 0, 0);
	if (!HeapTupleIsValid(tuple))
		elog(ERROR, "cache lookup failed for function %u", funcOid);
	proc = (Form_pg_proc) GETSTRUCT(tuple);

	functyptype = get_typtype(proc->prorettype);

	/* Disallow pseudotype result */
	/* except for TRIGGER, RECORD, or VOID */
	if (functyptype == TYPTYPE_PSEUDO)
	{
		/* we assume OPAQUE with no arguments means a trigger */
		if (proc->prorettype == TRIGGEROID ||
			(proc->prorettype == OPAQUEOID && proc->pronargs == 0))
			istrigger = true;
		else if (proc->prorettype != RECORDOID &&
				 proc->prorettype != VOIDOID)
			ereport(ERROR,
					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
					 errmsg("PL/Perl functions cannot return type %s",
							format_type_be(proc->prorettype))));
	}

	/* Disallow pseudotypes in arguments (either IN or OUT) */
	numargs = get_func_arg_info(tuple,
								&argtypes, &argnames, &argmodes);
	for (i = 0; i < numargs; i++)
	{
		if (get_typtype(argtypes[i]) == TYPTYPE_PSEUDO)
			ereport(ERROR,
					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
					 errmsg("PL/v8 functions cannot accept type %s",
							format_type_be(argtypes[i]))));
	}

	ReleaseSysCache(tuple);

	compile_plv8_function(funcOid, istrigger);
	
	/* the result of a validator is ignored */
	PG_RETURN_VOID();
}

//static Datum
//plv8_func_handler(PG_FUNCTION_ARGS)
//{
//	
//}

static plv8_info *
compile_plv8_function(Oid fn_oid, bool is_trigger)
{
	HeapTuple	procTup;
	Form_pg_proc	procStruct;
	char		internal_proname[NAMEDATALEN];
	plv8_info  *info = NULL;
	plv8_info_entry *hash_entry;
	bool		found;
	bool		isnull;
	Datum		procsrcdatum;
	char	   *proc_source;
	Oid		   *argtypes;
	char	  **argnames;
	char	   *argmodes;
	Oid			rettype;
//	Oid			functyptype;
//	Oid			retinput;
	MemoryContext oldcontext;
	int			i;

	procTup = SearchSysCache(PROCOID,
							 ObjectIdGetDatum(fn_oid),
							 0, 0, 0);
	if (!HeapTupleIsValid(procTup))
		elog(ERROR, "cache lookup failed for function %u", fn_oid);

	sprintf(internal_proname, "__PLv8_%u", fn_oid);

	hash_entry = (plv8_info_entry *) hash_search(plv8_info_hash, internal_proname, HASH_FIND, NULL);

	if (hash_entry)
	{
		bool		uptodate;

		info = hash_entry->info_data;

		uptodate = (info->fn_xmin == HeapTupleHeaderGetXmin(procTup->t_data) &&
					ItemPointerEquals(&info->fn_tid, &procTup->t_self));

		if (!uptodate)
		{
			pfree(info->prosrc);
			pfree(info->proname);
			info->function.Dispose();
			pfree(info);
			info = NULL;
			hash_search(plv8_info_hash, internal_proname, HASH_REMOVE, NULL);
		}
		else
		{
			ReleaseSysCache(procTup);
			return info;
		}
	}

	procStruct = (Form_pg_proc) GETSTRUCT(procTup);

	procsrcdatum = SysCacheGetAttr(PROCOID, procTup,
								   Anum_pg_proc_prosrc, &isnull);
	if (isnull)
		elog(ERROR, "null prosrc");

	proc_source = TextDatumGetCString(procsrcdatum);

	oldcontext = MemoryContextSwitchTo(TopMemoryContext);

	info = (plv8_info *) palloc0(sizeof(plv8_info));
	info->proname = pstrdup(NameStr(procStruct->proname));
	info->prosrc = pstrdup(proc_source);
	info->fn_xmin = HeapTupleHeaderGetXmin(procTup->t_data);
	info->fn_tid = procTup->t_self;
	info->nargs = get_func_arg_info(procTup,
									&argtypes, &argnames, &argmodes);

//	functyptype = get_typtype(procStruct->prorettype);
	rettype = procStruct->prorettype;

	ReleaseSysCache(procTup);

	for(i = 0; i < info->nargs; i++)
	{
		Oid		argtype = argtypes[i];
		char	argmode = argmodes ? argmodes[i] : PROARGMODE_IN;

		if (argmode != PROARGMODE_IN)
		{
			elog(ERROR, "argument %d must be IN parameter", i + 1);
		}

		plv8_fill_type(argtype, &info->argtypes[i], false, true);
	}

	plv8_fill_type(rettype, &info->rettype, true, false);

	info->function = plv8_create_function(internal_proname, info->nargs, (const char **) argnames, proc_source);
	hash_entry = (plv8_info_entry *) hash_search(plv8_info_hash, internal_proname, HASH_ENTER, &found);
	hash_entry->info_data = info;

	/* restore */
	MemoryContextSwitchTo(oldcontext);

	return info;
}

static v8::Persistent<v8::Function>
plv8_create_function(const char *proname, int proarglen, const char *proargs[], const char *prosrc)
{
	v8::HandleScope hscope;
	StringInfoData	args;
	StringInfoData	body;
	int				i;

	initStringInfo(&args);
	initStringInfo(&body);

	for(i = 0; i < proarglen; i++)
	{
		appendStringInfo(&args, "%s", proargs[i]);
		if (i != proarglen - 1)
		{
			appendStringInfoString(&args, ", ");
		}
	}

	/*
	 *	function <proname>(<arg1, ...>){
	 *		<prosrc>
	 *	}
	 *	<proname>
	 */
	appendStringInfo(&body, "function %s(%s){\n%s\n};\n%s", proname, args.data, prosrc, proname);
	
	v8::Handle<v8::Value> name = v8::String::New("compile");
	v8::Handle<v8::String> source = v8::String::New(body.data);
	v8::Context::Scope context_scope(global_context);
	v8::TryCatch try_catch;
	v8::Handle<v8::Script> script = v8::Script::Compile(source, name);

	if (script.IsEmpty())
	{
		plv8_elog_exception(ERROR, &try_catch);
	}

	v8::Handle<v8::Value> result = script->Run();
	if (result.IsEmpty())
	{
		plv8_elog_exception(ERROR, &try_catch);
	}

	if (!result->IsFunction())
	{
		elog(ERROR, "!result->IsFunction()");
	}

	return v8::Persistent<v8::Function>::New(v8::Handle<v8::Function>::Cast(result));
}

static Datum
plv8_call_jsfunction(FunctionCallInfo fcinfo)
{
	v8::HandleScope handle_scope;
	v8::Context::Scope context_scope(global_context);
	v8::Handle<v8::Value> args[FUNC_MAX_ARGS];
	int i;
	plv8_info	   *info = (plv8_info *) fcinfo->flinfo->fn_extra;
	plv8_type_info *rettype = &info->rettype;

	for(i = 0; i < info->nargs; i++)
	{
		plv8_type_info	   *type = &info->argtypes[i];

		if (type->category == TYPCATEGORY_ARRAY)
		{
			plv8_type_info	*elem = type->elem;

			args[i] = plv8_datum_to_jsarray(fcinfo->arg[i], fcinfo->argnull[i],
											elem->typid, elem->len,
											elem->byval, elem->align,
											&elem->fn_output);
		}
		else
		{
			args[i] = plv8_datum_to_jsval(fcinfo->arg[i], fcinfo->argnull[i],
										  type->typid, &type->fn_output);
		}
	}

	v8::Handle<v8::Value> result = info->function->Call(global_context->Global(), info->nargs, args);

	if (result.IsEmpty())
	{
		plv8_elog_exception(ERROR, &global_try_catch);
	}

	if (rettype->category == TYPCATEGORY_ARRAY)
	{
		plv8_type_info	   *elem = rettype->elem;

		return plv8_jsarray_to_datum(result,
									 &fcinfo->isnull,
									 elem->typid,
									 elem->ioparam,
									 elem->len,
									 elem->byval,
									 elem->align,
									 &elem->fn_input);
	}
	else
	{
		return plv8_jsval_to_datum(result, &fcinfo->isnull, rettype->typid, rettype->ioparam, &rettype->fn_input);
	}
/*
	else if (result->IsUndefined() || result->IsNull())
	{
		cstr_result = NULL;
		fcinfo->isnull = true;
		return (Datum) 0;
	}

	switch(rettype->typid)
	{
		case INT2OID:
			if (result->IsNumber())
				return Int16GetDatum((int16) (result->Int32Value()));
		case INT4OID:
			if (result->IsNumber())
				return Int32GetDatum((int32) (result->Int32Value()));
		case INT8OID:
			if (result->IsNumber())
				return Int64GetDatum((int64) (result->IntegerValue()));
		case FLOAT4OID:
			if (result->IsNumber())
				return Float4GetDatum((float4) (result->NumberValue()));
		case FLOAT8OID:
			if (result->IsNumber())
				return Float8GetDatum((float8) (result->NumberValue()));
		case BOOLOID:
			if (result->IsBoolean())
				return BoolGetDatum(result->BooleanValue());
		default:
			v8::String::Utf8Value str(result);
			cstr_result = const_cast<char *>(ToCString(str));
			cstr_result = plv8_tolocal(cstr_result);
			return InputFunctionCall(&rettype->fn_input, cstr_result, rettype->ioparam, -1);
	}
	elog(ERROR, "return type and returned value type unmatch");
	return (Datum) 0;
*/
}

static void
plv8_fill_type(Oid typid, plv8_type_info *type, bool forinput, bool foroutput)
{
	char		category;
	bool		ispreferred;

	type->typid = typid;
	get_type_category_preferred(typid, &category, &ispreferred);
	type->category = category;
	get_typlenbyvalalign(typid, &type->len, &type->byval, &type->align);

	if (forinput)
	{
		Oid		input_func;

		getTypeInputInfo(typid, &input_func, &type->ioparam);
		fmgr_info(input_func, &type->fn_input);
	}

	if (foroutput)
	{
		Oid		output_func;
		bool	isvarlen;

		getTypeOutputInfo(typid, &output_func, &isvarlen);
		fmgr_info(output_func, &type->fn_output);
	}

	if (category == TYPCATEGORY_ARRAY)
	{
		Oid			elemid = get_element_type(typid);
		plv8_type_info *elemtype = (plv8_type_info *) palloc0(sizeof(plv8_type_info));

		if (elemid == InvalidOid)
		{
			elog(ERROR, "type(%u):category == '%c' but elemid is invalid", typid, category);
		}
		
		plv8_fill_type(elemid, elemtype, forinput, foroutput);
		type->elem = elemtype;
	}
}

static Datum
plv8_jsarray_to_datum(v8::Handle<v8::Value> value,
					  bool *isnull,
					  Oid elemtype,
					  Oid elemioparam,
					  int16 elemlen,
					  bool elembyval,
					  char elemalign,
					  FmgrInfo *fn_eleminput)
{
	int			i, length;
	Datum	   *dvalues;
	bool	   *dnulls;
	int			ndims[0];
	int			lbs[] = {1};
	ArrayType  *array_type;

	if (value->IsUndefined() || value->IsNull())
	{
		*isnull = true;
		return (Datum ) 0;
	}

	if (!value->IsArray())
	{
		elog(ERROR, "value is not an Array");
	}

	v8::Handle<v8::Array> array(v8::Handle<v8::Array>::Cast(value));

	length = array->Length();
	dvalues = (Datum *) palloc(sizeof(Datum) * length);
	dnulls = (bool *) palloc0(sizeof(bool) * length);
	ndims[0] = length;
	for(i = 0; i < length; i++)
	{
		v8::Handle<v8::Value> el = array->Get(v8::Int32::New(i));
		dvalues[i] = plv8_jsval_to_datum(el, &dnulls[i], elemtype, elemioparam, fn_eleminput);
	}

	array_type = construct_md_array(dvalues, dnulls, 1, ndims, lbs, elemtype, elemlen, elembyval, elemalign);
	pfree(dvalues);
	pfree(dnulls);

	return (Datum) array_type;
}

static Datum
plv8_jsval_to_datum(v8::Handle<v8::Value> value,
					bool *isnull,
					Oid typid,
					Oid ioparam,
					FmgrInfo *fn_input)
{
	*isnull = false;
	if (value->IsUndefined() || value->IsNull())
	{
		*isnull = true;
		return (Datum) 0;
	}

	switch(typid)
	{
		case INT2OID:
			if (value->IsNumber())
				return Int16GetDatum((int16) (value->Int32Value()));
		case INT4OID:
			if (value->IsNumber())
				return Int32GetDatum((int32) (value->Int32Value()));
		case INT8OID:
			if (value->IsNumber())
				return Int64GetDatum((int64) (value->IntegerValue()));
		case FLOAT4OID:
			if (value->IsNumber())
				return Float4GetDatum((float4) (value->NumberValue()));
		case FLOAT8OID:
			if (value->IsNumber())
				return Float8GetDatum((float8) (value->NumberValue()));
		case BOOLOID:
			if (value->IsBoolean())
				return BoolGetDatum(value->BooleanValue());
		default:
			char *cstr;
			v8::String::Utf8Value str(value);
			cstr = const_cast<char *>(ToCString(str));
			cstr = plv8_tolocal(cstr);
			return InputFunctionCall(fn_input, cstr, ioparam, -1);
	}

	elog(ERROR, "datum type and js value type unmatch");
	return (Datum) 0;
}

static v8::Handle<v8::Value>
plv8_datum_to_jsarray(Datum datum,
					  bool isnull,
					  Oid elemtype,
					  int16 elemlen,
					  bool elembyval,
					  char elemalign,
					  FmgrInfo *fn_elemoutput)
{
	Datum	   *dvalues;
	bool	   *dnulls;
	int			nelems;
	int			i;

	if (isnull)
	{
		return v8::Null();
	}

	deconstruct_array((ArrayType *) datum, elemtype, elemlen, elembyval, elemalign,
					  &dvalues, &dnulls, &nelems);
	v8::Local<v8::Array>	result = v8::Array::New(nelems);
	for(i = 0; i < nelems; i++)
	{
		v8::Handle<v8::Value> val = plv8_datum_to_jsval(dvalues[i], dnulls[i], elemtype, fn_elemoutput);
		result->Set(v8::Int32::New(i), val);
	}

	return result;
}

static v8::Handle<v8::Value>
plv8_datum_to_jsval(Datum datum, bool isnull, Oid type, FmgrInfo *fn_output)
{
	char	   *value_text;

	if (isnull)
	{
		return v8::Null();
	}

	switch(type)
	{
		case INT2OID:
			return v8::Int32::New((int32_t) DatumGetInt16(datum));
		case INT4OID:
			return v8::Int32::New((int32_t) DatumGetInt32(datum));
		case INT8OID:
			return v8::Number::New((double) DatumGetInt64(datum));
		case FLOAT4OID:
			return v8::Number::New((double) DatumGetFloat4(datum));
		case FLOAT8OID:
			return v8::Number::New((double) DatumGetFloat8(datum));
		case BOOLOID:
			return v8::Boolean::New((bool) DatumGetBool(datum));
		default:
			value_text = OutputFunctionCall(fn_output, datum);
			value_text = plv8_toutf(value_text);
			return v8::String::New(value_text);
	}
}

static char *
plv8_toutf(char *cstr)
{
	int		db_encoding = GetDatabaseEncoding();

	if (db_encoding == PG_UTF8)
		return cstr;

	return (char *) pg_do_encoding_conversion(reinterpret_cast<unsigned char *>(cstr),
											  strlen(cstr), db_encoding, PG_UTF8);
}

static char *
plv8_tolocal(char *cstr)
{
	int		db_encoding = GetDatabaseEncoding();

	if (db_encoding == PG_UTF8)
		return cstr;

	return (char *) pg_do_encoding_conversion(reinterpret_cast<unsigned char *>(cstr),
											  strlen(cstr), PG_UTF8, db_encoding);
}

static void
plv8_elog_exception(int errcode, v8::TryCatch *try_catch)
{
	v8::String::Utf8Value exception(try_catch->Exception());
	const char *exception_string = ToCString(exception);

	elog(errcode, "%s", exception_string);
}

static const char *
ToCString(v8::String::Utf8Value &value)
{
	return *value ? *value : "<string conversion failed>";
}
