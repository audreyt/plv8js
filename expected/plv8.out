-- INSTALL
SET client_min_messages = warning;
\set ECHO none
RESET client_min_messages;
CREATE LANGUAGE plv8;
-- CREATE FUNCTION
CREATE FUNCTION plv8_test(keys text[], vals text[]) RETURNS text AS
$$
	var o = {};
	for (var i = 0; i < keys.length; i++)
		o[keys[i]] = vals[i];
	return JSON.stringify(o);
$$
LANGUAGE plv8 IMMUTABLE STRICT;
SELECT plv8_test(ARRAY['name', 'age'], ARRAY['Tom', '29']);
         plv8_test         
---------------------------
 {"name":"Tom","age":"29"}
(1 row)

CREATE FUNCTION unnamed_args(text[], text[]) RETURNS text[] AS
$$
	var array1 = arguments[0];
	var array2 = $2;
	return array1.concat(array2);
$$
LANGUAGE plv8 IMMUTABLE STRICT;
SELECT unnamed_args(ARRAY['A', 'B'], ARRAY['C', 'D']);
 unnamed_args 
--------------
 {A,B,C,D}
(1 row)

CREATE FUNCTION concat_strings(VARIADIC args text[]) RETURNS text AS
$$
    var result = "";
    for (var i = 0; i < args.length; i++)
    	if (args[i] != null)
	    	result += args[i];
    return result;
$$
LANGUAGE plv8 IMMUTABLE STRICT;
SELECT concat_strings('A', 'B', NULL, 'C');
 concat_strings 
----------------
 ABC
(1 row)

CREATE FUNCTION return_void() RETURNS void AS $$ $$ LANGUAGE plv8;
SELECT return_void();
 return_void 
-------------
 
(1 row)

CREATE FUNCTION return_null() RETURNS text AS $$ return null; $$ LANGUAGE plv8;
SELECT r, r IS NULL AS isnull FROM return_null() AS r;
 r | isnull 
---+--------
   | t
(1 row)

-- TYPE CONVERTIONS
CREATE FUNCTION int2_to_int4(x int2) RETURNS int4 AS $$ return x; $$ LANGUAGE plv8;
SELECT int2_to_int4(24::int2);
 int2_to_int4 
--------------
           24
(1 row)

CREATE FUNCTION int4_to_int2(x int4) RETURNS int2 AS $$ return x; $$ LANGUAGE plv8;
SELECT int4_to_int2(42);
 int4_to_int2 
--------------
           42
(1 row)

CREATE FUNCTION int4_to_int8(x int4) RETURNS int8 AS $$ return x; $$ LANGUAGE plv8;
SELECT int4_to_int8(48);
 int4_to_int8 
--------------
           48
(1 row)

CREATE FUNCTION int8_to_int4(x int8) RETURNS int4 AS $$ return x; $$ LANGUAGE plv8;
SELECT int8_to_int4(84);
 int8_to_int4 
--------------
           84
(1 row)

CREATE FUNCTION float8_to_numeric(x float8) RETURNS numeric AS $$ return x; $$ LANGUAGE plv8;
SELECT float8_to_numeric(1.5);
 float8_to_numeric 
-------------------
               1.5
(1 row)

CREATE FUNCTION numeric_to_int8(x numeric) RETURNS int8 AS $$ return x; $$ LANGUAGE plv8;
SELECT numeric_to_int8(1234.56);
 numeric_to_int8 
-----------------
            1234
(1 row)

CREATE FUNCTION int4_to_text(x int4) RETURNS text AS $$ return x; $$ LANGUAGE plv8;
SELECT int4_to_text(123);
 int4_to_text 
--------------
 123
(1 row)

CREATE FUNCTION text_to_int4(x text) RETURNS int4 AS $$ return x; $$ LANGUAGE plv8;
SELECT text_to_int4('123');
 text_to_int4 
--------------
          123
(1 row)

SELECT text_to_int4('abc'); -- error
ERROR:  invalid input syntax for integer: "abc"
CREATE FUNCTION int4array_to_textarray(x int4[]) RETURNS text[] AS $$ return x; $$ LANGUAGE plv8;
SELECT int4array_to_textarray(ARRAY[123, 456]::int4[]);
 int4array_to_textarray 
------------------------
 {123,456}
(1 row)

CREATE FUNCTION textarray_to_int4array(x text[]) RETURNS int4[] AS $$ return x; $$ LANGUAGE plv8;
SELECT textarray_to_int4array(ARRAY['123', '456']::text[]);
 textarray_to_int4array 
------------------------
 {123,456}
(1 row)

CREATE FUNCTION timestamptz_to_text(t timestamptz) RETURNS text AS $$ return t.toUTCString() $$ LANGUAGE plv8;
SELECT timestamptz_to_text('23 Dec 2010 12:34:56 GMT');
      timestamptz_to_text      
-------------------------------
 Thu, 23 Dec 2010 12:34:56 GMT
(1 row)

CREATE FUNCTION text_to_timestamptz(t text) RETURNS timestamptz AS $$ return new Date(t) $$ LANGUAGE plv8;
SELECT text_to_timestamptz('23 Dec 2010 12:34:56 GMT') AT TIME ZONE 'GMT';
         timezone         
--------------------------
 Thu Dec 23 12:34:56 2010
(1 row)

CREATE FUNCTION date_to_text(t date) RETURNS text AS $$ return t.toUTCString() $$ LANGUAGE plv8;
SELECT date_to_text('23 Dec 2010');
         date_to_text          
-------------------------------
 Thu, 23 Dec 2010 00:00:00 GMT
(1 row)

CREATE FUNCTION text_to_date(t text) RETURNS date AS $$ return new Date(t) $$ LANGUAGE plv8;
SELECT text_to_date('23 Dec 2010 GMT');
 text_to_date 
--------------
 12-23-2010
(1 row)

CREATE FUNCTION oidfn(id oid) RETURNS oid AS $$ return id $$ LANGUAGE plv8;
SELECT oidfn('pg_catalog.pg_class'::regclass);
 oidfn 
-------
  1259
(1 row)

-- RECORD TYPES
CREATE TYPE rec AS (i integer, t text);
CREATE FUNCTION scalar_to_record(i integer, t text) RETURNS rec AS
$$
	return { "i": i, "t": t };
$$
LANGUAGE plv8;
SELECT scalar_to_record(1, 'a');
 scalar_to_record 
------------------
 (1,a)
(1 row)

CREATE FUNCTION record_to_text(x rec) RETURNS text AS
$$
	return JSON.stringify(x);
$$
LANGUAGE plv8;
SELECT record_to_text('(1,a)'::rec);
 record_to_text  
-----------------
 {"i":1,"t":"a"}
(1 row)

CREATE FUNCTION set_of_records() RETURNS SETOF rec AS
$$
	yield( { "i": 1, "t": "a" } );
	yield( { "i": 2, "t": "b" } );
	yield( { "i": 3, "t": "c" } );
$$
LANGUAGE plv8;
SELECT * FROM set_of_records();
 i | t 
---+---
 1 | a
 2 | b
 3 | c
(3 rows)

CREATE FUNCTION set_of_integers() RETURNS SETOF integer AS
$$
	yield( 1 );
	yield( 2 );
	yield( 3 );
$$
LANGUAGE plv8;
SELECT * FROM set_of_integers();
ERROR:  SETOF a scalar type is not implemented yet
-- print()
CREATE FUNCTION test_print(arg text) RETURNS void AS
$$
	print(NOTICE, 'args =', arg);
	print(WARNING, 'warning');
	print(20, 'ERROR is not allowed');
$$
LANGUAGE plv8;
SELECT test_print('ABC');
NOTICE:  args = ABC
WARNING:  warning
ERROR:  invalid error level for print()
DETAIL:  test_print() LINE 4: 	print(20, 'ERROR is not allowed');
-- executeSql()
CREATE TABLE test_tbl (i integer, s text);
CREATE FUNCTION test_sql() RETURNS integer AS
$$
	var rows = executeSql("SELECT i, 's' || i AS s FROM generate_series(1, 4) AS t(i)");
	for (var r = 0; r < rows.length; r++)
	{
		var result = executeSql("INSERT INTO test_tbl VALUES(" + rows[r].i + ",'" + rows[r].s + "')");
		print(NOTICE, JSON.stringify(rows[r]), result);
	}
	return rows.length;
$$
LANGUAGE plv8;
SELECT test_sql();
NOTICE:  {"i":1,"s":"s1"} 1
NOTICE:  {"i":2,"s":"s2"} 1
NOTICE:  {"i":3,"s":"s3"} 1
NOTICE:  {"i":4,"s":"s4"} 1
 test_sql 
----------
        4
(1 row)

SELECT * FROM test_tbl;
 i | s  
---+----
 1 | s1
 2 | s2
 3 | s3
 4 | s4
(4 rows)

CREATE FUNCTION test_sql_error() RETURNS void AS $$ executeSql("ERROR") $$ LANGUAGE plv8;
SELECT test_sql_error();
ERROR:  syntax error at or near "ERROR"
DETAIL:  test_sql_error() LINE 1:  executeSql("ERROR") 
-- TRIGGER
CREATE FUNCTION test_trigger() RETURNS trigger AS
$$
	print(NOTICE, "NEW = ", JSON.stringify(NEW));
	print(NOTICE, "OLD = ", JSON.stringify(OLD));
	print(NOTICE, "TG_OP = ", TG_OP);
	print(NOTICE, "TG_ARGV = ", TG_ARGV);
$$
LANGUAGE "plv8";
CREATE TRIGGER test_trigger
  BEFORE INSERT OR UPDATE OR DELETE
  ON test_tbl FOR EACH ROW
  EXECUTE PROCEDURE test_trigger('foo', 'bar');
INSERT INTO test_tbl VALUES(100, 'ABC');
NOTICE:  NEW =  {"i":100,"s":"ABC"}
NOTICE:  OLD =  undefined
NOTICE:  TG_OP =  INSERT
NOTICE:  TG_ARGV =  foo,bar
UPDATE test_tbl SET i = 101, s = 'DEF' WHERE i = 1;
NOTICE:  NEW =  {"i":101,"s":"DEF"}
NOTICE:  OLD =  {"i":1,"s":"s1"}
NOTICE:  TG_OP =  UPDATE
NOTICE:  TG_ARGV =  foo,bar
DELETE FROM test_tbl WHERE i >= 100;
NOTICE:  NEW =  undefined
NOTICE:  OLD =  {"i":100,"s":"ABC"}
NOTICE:  TG_OP =  DELETE
NOTICE:  TG_ARGV =  foo,bar
NOTICE:  NEW =  undefined
NOTICE:  OLD =  {"i":101,"s":"DEF"}
NOTICE:  TG_OP =  DELETE
NOTICE:  TG_ARGV =  foo,bar
SELECT * FROM test_tbl;
 i | s  
---+----
 2 | s2
 3 | s3
 4 | s4
(3 rows)

-- ERRORS
CREATE FUNCTION syntax_error() RETURNS text AS '@' LANGUAGE plv8;
ERROR:  SyntaxError: Unexpected token ILLEGAL
DETAIL:  syntax_error() LINE 1: @
CREATE FUNCTION reference_error() RETURNS text AS 'not_defined' LANGUAGE plv8;
SELECT reference_error();
ERROR:  ReferenceError: not_defined is not defined
DETAIL:  reference_error() LINE 1: not_defined
CREATE FUNCTION throw() RETURNS void AS $$throw new Error("an error");$$ LANGUAGE plv8;
SELECT throw();
ERROR:  Error: an error
DETAIL:  throw() LINE 1: throw new Error("an error");
